### PR TITLE
fixes

### PR DESCRIPTION
**Pull request by imnotbob**
Un declared local variables
Incorrect recursive calls
Not tracking subscriptions properly
general cleanups, code optimizations

### CODE REVIEW

File: SleepNumberController_App.groovy [hunk 1]
```diff
@@ -58,7 +58,7 @@ preferences {
   page name: "createBedPage"
   page name: "diagnosticsPage"
 }
-
+/*
```
Comment (rvrolyk) [inline SleepNumberController_App.groovy:61]: this is, iirc, required to handle the pause/resume button at the top of the app.

File: SleepNumberController_App.groovy [hunk 2]
```diff
@@ -304,18 +305,18 @@ def scheduledRefreshChildDevices() {
...
   if (settings.modes && !settings.modes.contains(location.mode)) {
     return
   }
   getBedData()
   updateLabel()
 }
-
-def refreshChildDevices(ignored, ignoredDevId) {
+/*
+void refreshChildDevices(ignored, ignoredDevId) {
```
Comment (rvrolyk) [inline SleepNumberController_App.groovy:317]: This is required so the driver can call this method (it sends two arguments so the no argument variant won't work).

File: SleepNumberController_App.groovy [hunk 3]
```diff
@@ -767,162 +772,164 @@ def processBedData(responseData) {
...
   def underbedLightData = [:]
 
-  def deviceTypes = getBedDeviceTypes()
+  List deviceTypes = getBedDeviceTypes()
 
   for (def device : getBedDevices()) {
-    if (!outletData.get(device.getState().bedId)) {
-      outletData[device.getState().bedId] = [:]
-      underbedLightData[device.getState().bedId] = [:]
+    String bedId=device.getState().bedId.toString()
```
Comment (rvrolyk) [inline SleepNumberController_App.groovy:778]: nit: space before and after =

**Review by rvrolyk — CHANGES_REQUESTED** [review]

**Comment by rvrolyk** [comment]
also thanks for the fixes!  wanted to ask about the typing since I thought groovy was ok w/ untyped; is this a style thing or is there some other improvement using specific types?

File: SleepNumberController_App.groovy (hunk 3 above)
Comment (imnotbob) [reply SleepNumberController_App.groovy:778]: If you like adjust to to your style conventions

File: SleepNumberController_App.groovy (hunk 2 above)
Comment (imnotbob) [reply SleepNumberController_App.groovy:317]: ok, you can reject that block

File: SleepNumberController_App.groovy (hunk 1 above)
Comment (imnotbob) [reply SleepNumberController_App.groovy:61]: I don't see any references.  It looks like a ST thing to me.

**Comment by imnotbob** [comment]
On typing, it reduces code size to remove box/unbox in the code.  Similar for the reduction in device state calls, it avoids a lot fo runtime in HE.  This is all stuff I learned optimizing webcore, homebridge, echo speaks, OpenWeatherMap

File: SleepNumberController_App.groovy (hunk 1 above)
Comment (rvrolyk) [reply SleepNumberController_App.groovy:61]: Definitely a thing ... I just commented it out locally and tried clicking the pause button and HE emits this error:

2020-12-17 03:36:04.092 pm errorgroovy.lang.MissingMethodException: No signature of method: user_app_rvrolyk_Sleep_Number_Controller_898.appButtonHandler() is applicable for argument types: (java.lang.String) values: [pause] (appButtonHandler)

**Review by rvrolyk — APPROVED** [review]

**Pull request merged**